### ISSUE TITLE
Deriving Read instance causes unsoundness

### ISSUE THREAD

**Issue by nikivazou** [ucsd-progsys/liquidhaskell#743]
The [following example](https://github.com/ucsd-progsys/liquidhaskell/blob/8fe9dfb0f1e194f51809ee428113816dd9b83101/tests/todo/DerivingRead.hs) is SAFE while should be unsafe. 

```
{-@ checkNat :: Nat -> Int @-} 
checkNat :: Int -> Int 
checkNat x = x

unsound :: Int
unsound = checkNat (-1)


data TestBS = TestBS Int deriving (Read)
```
Current workarounds
 -  add `--trust-internals` flag to ignore the deriving instances
 -  delete the `deriving (Read)` instance

**Comment by ranjitjhala** [comment]
This is a long standing problem, do you have any idea what the cause is? Does the read instance create some top-level literals?

**Comment by ranjitjhala** [comment]
@nikivazou why don't we just make `--trust-internals` the default behavior and *not* an option ?

**Comment by nikivazou** [comment]
I did not look more into it. Just minimized the code and filed the issue. It should be a top level binder. 
Yes lets make trust internals default! 

Also, suggesting to make the totality default!

**Comment by ranjitjhala** [comment]
I looked at the output core, and basically there are *many* little top-level binders added which are not "called" and so probably get the type "False" (as they are not explicitly exported...)

**Comment by ranjitjhala** [comment]
totality default...? hmm...seems brave... :)

**Comment by nikivazou** [comment]
I am exporting everything, so even if they are not called they should be refined to true! 

I think totality is more common and easier to prove than termination. And totality and termination are nice default checks: if your code passes LiquidHaskell then it is total!

**Comment by ranjitjhala** [comment]
ok, will do.

**Comment by nikivazou** [comment]
Or we can do it after the deadline:)

**Comment by ranjitjhala** [comment]
Hopefully fixed by https://github.com/ucsd-progsys/liquidhaskell/pull/749

**Comment by ranjitjhala** [comment]
Unfortunately, the above doesn't work, as it breaks all our class measure stuff (all those *also* generate binders that look like "GHC internals" or are "derived" names...)

In the case of `Read` there are some odd recursive definitions, namely

`Bob.$fReadTestBS` depends on `$creadListPrec` which itself depends on `Bob.$fReadTestBS`,
yielding the `False` . I'm guessing _something_ here needs to be contra-variant?


```
 Bob.$fReadTestBS =
   (((GHC.Read.D:Read @ Bob.TestBS $creadsPrec) $creadList)
      $creadPrec)
     $creadListPrec;
 $creadListPrec [Occ=LoopBreaker]
   :: Text.ParserCombinators.ReadPrec.ReadPrec [Bob.TestBS]
 [LclId, Str=DmdType]
 $creadListPrec =
   GHC.Read.readListPrecDefault @ Bob.TestBS Bob.$fReadTestBS;`

**Comment by ranjitjhala** [comment]
Ok, so looking at the definitions (links below) I see that:

```haskell
newtype ReadPrec a = P (Prec -> ReadP a)
newtype ReadP a = R (forall b . (a -> P b) -> P b)
```

Does that mean the `a` is contra-variant? (Some fancy recursion going on in there :-))
 

[ReadPrec](http://hackage.haskell.org/package/base-4.9.0.0/docs/src/Text.ParserCombinators.ReadPrec.html#ReadPrec)

[ReadP](http://hackage.haskell.org/package/base-4.9.0.0/docs/src/Text.ParserCombinators.ReadP.html#ReadP)

**Comment by ranjitjhala** [comment]
The problem is not contra variance IMO, but the odd recursion. Here's a simplified version of the generated code -- it is **rejected** by the termination checker. 

**Question** do you somehow disable the termination checker for DERIVED functions? (Otherwise, why isn't complaining about this weird mutual recursion?)
 

```haskell
module Bob (bar) where

data Foo a = FooCon a
data Dict = DictCon

{-@ foo :: {v:Foo Int | false} @-}
foo = undefined :: Foo Int

{-@ mkDict :: Foo Int -> Dict @-}
mkDict :: Foo Int -> Dict
mkDict _ = DictCon

dict      = mkDict dictList
dictList  = readListPrecDefault dict

{-@ readListPrecDefault :: Dict -> Foo Int @-}
readListPrecDefault :: Dict -> Foo Int
readListPrecDefault = undefined

{-@ bar :: Nat @-}
bar :: Int
bar = 2 - 10
```

**Comment by ranjitjhala** [comment]
Reading the code, the termination checker *is* in-fact disabled for derived methods. I suppose it has to be because there is no easy way to attach annotations like "lazy" or termination-metrics etc. to them. Instead the fix in the PR is to ensure that the top-level refinements are `True` so they don't "pollute" the environment with an absurdity.

**Comment by ranjitjhala** [comment]
Fixed by PR #749
